[FLINK-2961] [table] Add support for basic type Date in Table API

twalthr opened the pull request:
Currently, the basic type Date is not implemented in the Table API. In order to have a mapping of the most important ANSI SQL types for FLINK-2099. It makes sense to add support for Date to represent date, time and timestamps of milliseconds precision. This PR implements FLINK-2961.

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/codegen/ExpressionCodeGenerator.scala:
```diff
@@ -202,6 +210,23 @@ abstract class ExpressionCodeGenerator[R](
+
+        if (nullCheck) {
+          s"""
+            |val $nullTerm = false
```
Aren't you generating Java code here? This should not work with `val`. I assume that you don't have a test case where you generate this code. Otherwise it should have failed. I guess it would be good to add a test case for it.

aljoscha replied:
These are not tested anywhere. We should maybe add tests that also verify the Code Generator when `nullCheck` is `true`.

twalthr replied:
Good point. `val` occurs in all literal code generation lines. I copy pasted the bug.

aljoscha commented:
So all timestamps have to be in UTC timezone. Isn't this a bit limiting? Could it per default take the system timezone and allow to override the timezone in a cast statement?

twalthr commented:
I also also thought about that but I'm pretty sure that this introduces issues difficult to debug. I think the most important thing is that the conversion (from string to date and back) is always constant. Internally date is always UTC anyway. Introducing a `.convert()` function for that might be better. My goal was primarily to have this data type in the Table API.

twalthr commented:
It seems that `nullCheck` has never been `true`. The code generator had lots of bugs :D

twalthr commented:
I think the best solution (also for future features like the SQL API) is a TableConfig, where TimeZone, Locales, and also the `nullCheck` can be configured. Once #1237 is merged (hopefully soon), I will implement that into the `TableEnvironment`. What do you think?

twalthr commented:
I have updated my code again. Internally I already implemented a TableConfig that defines the TimeZone and NullCheck. Once #1237 is merged, I will make the config available through TableEnvironment as part of FLINK-3068. What do you think?

aljoscha commented:
Looks good now.

twalthr commented:
I will merge this later today, if nobody has objections.
